tonic/functional/to_averaged_timesurface.py

```diff
@@ -1,146 +1,132 @@
 import numpy as np
-
-
-def findCell(x, y, bounds):
-
-    # build point
-    points = np.array([[x, y]])
-
-    # check for each event if all coordinates are in bounds
-    allInBounds = points[:, 0] >= bounds[:, None, 0]
-    allInBounds &= points[:, 1] >= bounds[:, None, 1]
-    allInBounds &= points[:, 0] < bounds[:, None, 2]
-    allInBounds &= points[:, 1] < bounds[:, None, 3]
-
-    # now find out the positions of all nonzero values
-    nz = np.nonzero(allInBounds)
-
-    # initialize the result with all nan
-    r = np.full(points.shape[0], np.nan)
-
-    # use nz[1] to index event position and nz[0] to tell which cell the event belongs to
-    r[nz[1]] = nz[0]
-    return int(r)
-
-
+from joblib import Parallel, delayed, cpu_count
+import math
+
+def _gen_pixel_to_cell_mapper(w, h, K, px_dtype):
+  matrix = np.zeros((h, w), dtype=px_dtype)
+  for i in range(h):
+    for j in range(w):
+      # Since each frame axis is divided in cells of width _K, the cell index related to that axis can be computed by integer division x//_K.
+      cell_y, cell_x = i//K, j//K
+      # Getting the index associated to cell coordinates.
+      matrix[i,j] = cell_y*math.ceil(w/K) + cell_x
+  return matrix
+  
+def _get_cell_mems(events, n_cells, n_pols, pixel_to_cell_mapper):
+  cell_mems = [[[] for c in range(n_cells)] for p in range(n_pols)]
+  for event in events:
+    cell_mems[max(event['p'], 0)][pixel_to_cell_mapper[event['y'], event['x']]].append(event)
+  return cell_mems[:]
+
+def _bsearch_window(t_start, cell_mem):
+  l, r, start = 0, len(cell_mem)-1, 0
+  while l<=r:
+    m = (l+r)//2
+    if cell_mem[m]['t'] > t_start:
+      r, start = m-1, m
+    elif cell_mem[m]['t'] < t_start:
+      l = m+1
+    else:
+      start = m
+      break
+  return start
+
+def _get_time_surface(event, cell_mem, radius, temporal_window, tau):
+  ts = np.zeros((2*radius+1, 2*radius+1), dtype=np.float32)
+  # Getting starting timestamp.
+  t_start = max(0, event['t'] - temporal_window)
+  check_coords = lambda x, y: x>=0 and y>=0 and x <= 2*radius and y <= 2*radius
+  # Getting start timestamp of temporal window.
+  start = _bsearch_window(t_start=t_start, cell_mem=cell_mem)
+  for mem_event in cell_mem[start:]:
+    # Getting event coordinates in the neighbourhood.
+    x, y = (event['x'] - mem_event['x']) + radius, (event['y'] - mem_event['y']) + radius
+    # Check if event is in neighourbhood and, if so, adding it to the time surface.
+    if check_coords(x, y):
+      ts[y,x] += np.exp(-(event['t'] - mem_event['t']).astype(np.float32)/tau)
+  return ts
```

Comment:
would it be possible to use https://github.com/neuromorphs/tonic/blob/develop/tonic/functional/to_timesurface.py for this? Potentially adapting that file if need be

Reply:
I will look into it :) I was thinking of determining the cell buffer slice with `_bsearch_window`, transforming it in a numpy array and passing it to `to_time_surface.py`. The only thing that I was thinking about is that, in the original HOTS paper, the time surface was computed by considering only the closest timestamp to the event considered. 
![image](https://user-images.githubusercontent.com/45739782/160040198-f9363af8-c684-4535-9ad4-1275c6234a31.png)
Instead, in HATS we need to accumulate the timesurfaces considering all the events in the time window. The time surface transform, hence, should be called on each of these event, which is a very large number of times. Am I right? My concerns are two: 
- number of times the time surface function is called. 
- time surface would need to compute the closest timestamp to the event each time, unless we pass a single-event buffer to the function.

However, this is only a gut feeling but I wanted to know you opinion about it :)